posts/2024-05-22-multi-dim-arrays.qmd

```diff
@@ -0,0 +1,283 @@
+samples, and as a result these dimensions should be displayed horizontally and vertically, respectively.  Most formats for
+storing "natural" images store data such that the "horizontal axis" / rows have a smaller stride than the "vertical axis" /
+columns. Note: while rows have smaller stride than columns, it is common for rows not to have stride 1, for example when using
+"interleaved" color components, the "color" dimension will have a stride of 1. Biomedical images do not typically have
+"horizontal" or "vertical" dimensions, but may have other semantics (e.g., anatomical, or related to the imaging system).
+
+**Recommendation:** Software and storage formats should be explicit about any semantics that are attached to dimensions and
+not use stride as a proxy for or to imply semantics. Documentation may specify semantics in terms of fastest/slowest dimension,
+but must explicitly communicate that fact. 
+
+For images, C- and F-order will historically disagree on the ordering of an array's indexes and dimensions, but will store the
+arrays the same way when flattened. This is because the convention for natural images historically 
+associates dimension semantics to the memory layout (the fastest dimension has horizontal semantics).
+
+## Coordinate data
+
+Coordinate data are data that refer to "locations" of the multidimensional array. They may be discrete, 
+and refer to specific samples, or continuous, referring to points "in-between" array locations.
+
+::: {.callout-tip collapse="true" appearance="minimal"}
+# Coordinate data examples
+
+* Point annotations
+    * "Structure `A` is located at point `(x,y)`"
+* Bounding boxes / ROIs
+    * "Crop the image to bounding box `(min_x, min_y, width, height)`
+* Other collections of points
+    * "My neuron skeleton consists of points `[[z0, y0, x0], ..., [zN, yN, xN]]`
+
+:::
+
+
+
+
+### cartesian coordinates
+
+In contrast to the matrix row/column index convention, cartesian
+coordinates label the horizontal and vertical dimensions `x` and `y`,
+respectively. Referencing positions in the 2D plane is done using
+ordered two-tuples `(x,y)`, where `x` is conventionally the left-index
+and `y` is the right-index. Using cartesian coordinates, varying
+the left dimensions varies horizontal position, and varying the right
+dimension varies the vertical position.
+
+Applications and workflows that make use of image geometry most commonly use cartesian coordinates.
+
+
+## References
+
+1) [hdf5 dataspaces](https://docs.hdfgroup.org/hdf5/develop/_h5_s__u_g.html#sec_dataspace)
+2) [zarr arrays](https://zarr-specs.readthedocs.io/en/latest/v2/v2.0.html#arrays)
+3) [nrrd axis ordering](https://teem.sourceforge.net/nrrd/format.html#general.4)
+4) [n5 ordering discussion](https://github.com/saalfeldlab/n5/issues/31)
+5) [multi-dimensional arrays in vigra](http://ukoethe.github.io/vigra/doc-release/vigranumpy/index.html#more-on-the-motivation-and-use-of-axistags)
+
+## Appendix
+
+### Programming languages
```

Comment:
I might just scrap the whole table, since I'm less and less sure what the added value is.  The scope of this may now be such that anyone that this article is useful for already knows what's in that table and more details of the type you're pointing out...

Reply:
I think it might be useful for someone moving data from NumPy to imglib2 to see what the _default_ order is of those two libraries.